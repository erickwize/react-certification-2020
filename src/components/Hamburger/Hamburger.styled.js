import styled from 'styled-components'

const StyledHamburger = styled.div.attrs({
  className: 'hamburger',
})`
  cursor: pointer;
  z-index: 10;
  padding: 0.8em;
  display: flex;
  flex-direction: column;
  justify-content: center;

  span {
    width: ${({ toggle }) => (toggle ? '1.56em' : '1.44em')};
    height: 0.19em;
    background-color: ${({ toggle }) => (toggle ? '#FFFFFF' : '#2d314d')};
    border-radius: 1em;
    transform-origin: 0.1em 0.1em;
    margin-bottom: 0.3em;
    z-index: 10;
    transition-property: transform;
    transition-duration: 0.1s;
    transition-timing-function: ease;

    &:nth-child(1) {
      transform: ${({ toggle }) => (toggle ? 'rotate(45deg)' : 'rotate(0deg)')};
    }
    &:nth-child(2) {
      opacity: ${({ toggle }) => (toggle ? '0' : '1')};
    }
    &:last-child {
      transform: ${({ toggle }) => (toggle ? 'rotate(-45deg)' : 'rotate(00deg)')};
    }
  }
`
export default StyledHamburger
