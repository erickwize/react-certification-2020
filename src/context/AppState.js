/* global gapi*/
/* eslint-disable react/prop-types */
import React, { useReducer } from "react";
import AppContext from "./appContext";
import AppReducer from "./appReducer";
import validateItems from "../utils/validateItems";
import initGapi from "../utils/initGapi";
import { GET_RESULT_VIDEOS, GET_RELATED_VIDEOS, SET_LOADING } from "./types";

const AppState = (props) => {
  const initialState = {
    searchText: "Moonwalk",
    resultVideos: [],
    selectedVideo: {},
    relatedVideos: [],
    showPlayer: false,
    loading: true,
  };

  const [state, dispatch] = useReducer(AppReducer, initialState);

  // HANDLE INIT GAPI
  const handleInitGapi = async () => {
    initGapi();
    setTimeout(() => {
      getResultVideos(state.searchText);
    }, 1000);
  };

  // GET RESULT VIDEOS
  const getResultVideos = async (query) => {
    setLoading();
    let items;

    try {
      const response = await gapi.client.youtube.search.list({
        part: ["snippet"],
        maxResults: 50,
        q: query,
        type: ["video"],
      });
      items = validateItems(response.result.items);
    } catch (error) {
      console.error("getResultVideos: Something went wrong... ", error);
    }

    dispatch({
      type: GET_RESULT_VIDEOS,
      payload: { query, items },
    });
  };

  // GET RELATED VIDEOS
  const getRelatedVideos = async (video) => {
    setLoading();
    let items;

    try {
      const response = await gapi.client.youtube.search.list({
        part: ["snippet"],
        maxResults: 50,
        type: ["video"],
        relatedToVideoId: video.id,
      });
      items = validateItems(response.result.items);
    } catch (error) {
      console.error("getRelatedVideos: Something went wrong... ", error);
    }

    dispatch({
      type: GET_RELATED_VIDEOS,
      payload: { video, items },
    });
  };

  // SET LOADING
  const setLoading = () => {
    dispatch({ type: SET_LOADING });
  };

  return (
    <AppContext.Provider
      value={{
        searchText: state.searchText,
        resultVideos: state.resultVideos,
        selectedVideo: state.selectedVideo,
        relatedVideos: state.relatedVideos,
        showPlayer: state.showPlayer,
        loading: state.loading,
        handleInitGapi,
        getResultVideos,
        getRelatedVideos,
        setLoading,
      }}
    >
      {props.children}
    </AppContext.Provider>
  );
};

export default AppState;
