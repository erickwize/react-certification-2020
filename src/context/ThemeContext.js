import React, { createContext, useState } from 'react'

const ThemeContext = createContext()

const THEMES = {
	light: {
		background: '#eeeeee',
		color: '#555555',
		colorTitle: '#333333',
		headerBackground: '#dddddd',
		cardBackground: '#e6e6e6',
	},
	dark: {
		background: '#33333d',
		color: '#b7b7ba',
		colorTitle: '#f9f9f9',
		headerBackground: '#424251',
		cardBackground: '#1a1a1a',
	},
}
function ThemeContextProvider({ children }) {
	const [theme, setTheme] = useState('dark')
	const defaultContext = {
		theme,
		toggle: () => setTheme(theme === 'light' ? 'dark' : 'light'),
		...THEMES[theme],
	}

	return <ThemeContext.Provider value={defaultContext}>{children}</ThemeContext.Provider>
}

export { ThemeContext, ThemeContextProvider }
