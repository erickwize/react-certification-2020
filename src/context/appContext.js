import { createContext, useContext } from "react";

const appContext = createContext();

export const useAppContext = () => {
  const context = useContext(appContext);
  if (context === undefined) {
    throw new Error(
      "useAppContext must be used within AppState component tree"
    );
  }
  return context;
};

export default appContext;
