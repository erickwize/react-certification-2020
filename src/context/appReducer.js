import { GET_RESULT_VIDEOS, GET_RELATED_VIDEOS, SET_LOADING } from "./types";

export default (state, action) => {
  switch (action.type) {
    case GET_RESULT_VIDEOS:
      return {
        ...state,
        searchText: action.payload.query,
        resultVideos: action.payload.items,
        showPlayer: false,
        loading: false,
      };
    case GET_RELATED_VIDEOS:
      return {
        ...state,
        selectedVideo: action.payload.video,
        relatedVideos: action.payload.items,
        showPlayer: true,
        loading: false,
      };
    case SET_LOADING:
      return {
        ...state,
        loading: true,
      };
    default:
      return state;
  }
};
