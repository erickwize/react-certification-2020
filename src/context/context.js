// import Dexie from 'dexie'
import React, { useEffect, createContext, useReducer, memo } from 'react'
import { useFetch } from '../utils/hooks/useFetch'

export const Context = createContext(null)

export const ContextReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_THEME':
      return {
        ...state,
        theme: !state.theme,
      }
    case 'OPEN_MENU':
      return {
        ...state,
        menu: action.payload,
      }
    case 'GET_DATA':
      return {
        ...state,
        data: action.payload,
      }
    case 'SET_VIEW':
      return {
        ...state,
        view: action.payload,
      }
    case 'SET_TARGET':
      return {
        ...state,
        target: action.payload,
      }
    default:
      return state
  }
}

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data, isLoading } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])

  return isLoading ? (
    <div> Loading...</div>
  ) : (
    <Context.Provider value={{ state, dispatch }}> {children} </Context.Provider>
  )
}
export default memo(ContextProvider)
