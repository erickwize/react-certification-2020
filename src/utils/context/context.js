import { createContext } from 'react';

export const Context = createContext({
  queryToSearch: null,
  setQueryFromSearchInput: null,
});
