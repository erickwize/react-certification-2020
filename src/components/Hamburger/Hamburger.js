import React, { useContext, useCallback } from 'react'
import { Context } from '../../context/context'
import StyledHamburger from './Hamburger.styled'

const Hamburger = React.memo(() => {
  const { state, dispatch } = useContext(Context)
  const openMenu = useCallback(() => {
    dispatch({ type: 'OPEN_MENU', payload: true })
  }, [dispatch])

  return (
    <StyledHamburger key="Hamburger" onClick={openMenu} toggle={state.menu}>
      <span />
      <span />
      <span />
    </StyledHamburger>
  )
})
export default Hamburger
