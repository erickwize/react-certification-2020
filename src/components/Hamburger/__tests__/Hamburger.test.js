import React from 'react'
import renderer from 'react-test-renderer'
import { Context } from '../../../context/context'
import Hamburger from '../Hamburger'

const state = {
  theme: false,
}

describe('Testing Hamburger component', () => {
  test('renders Hamburger without crashing', async () => {
    const rendered = renderer.create(
      <Context.Provider value={{ state }}>
        <Hamburger />
      </Context.Provider>
    )
    expect(rendered).toMatchSnapshot()
  })
})
