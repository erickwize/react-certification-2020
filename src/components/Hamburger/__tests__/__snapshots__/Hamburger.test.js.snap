// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Hamburger component renders Hamburger without crashing 1`] = `
<div
  className="sc-bdvvaa iriGUn hamburger"
  onClick={[Function]}
>
  <span />
  <span />
  <span />
</div>
`;
